config_api.go

```diff
@@ -25,41 +27,57 @@ type ClientConfig struct {
 	S3URLPrefix    string
 }
 
-func (c *ClientConfig) populateClientBinaries(ctx context.Context, bucket pail.Bucket, prefix string) error {
-	if !strings.HasSuffix(prefix, "/") {
-		prefix += "/"
-	}
+func (c *ClientConfig) populateClientBinaries(ctx context.Context, s3URLPrefix string) {
+	client := utility.GetHTTPClient()
+	defer utility.PutHTTPClient(client)
 
-	iter, err := bucket.List(ctx, prefix)
-	if err != nil {
-		return errors.Wrap(err, "listing client bucket")
-	}
-	for iter.Next(ctx) {
-		item := strings.TrimPrefix(iter.Item().Name(), prefix)
-		// The item is expected to be of the form <os>_<arch>/evergreen{.exe}
-		name := strings.Split(item, "/")
-		// Skip files that either don't conform to the expected form or aren't the evergreen executable (e.g. .signed).
-		if len(name) != 2 || !strings.Contains(name[1], "evergreen") {
+	// We assume that all valid OS/arch combinations listed in Evergreen
+	// constants are supported platforms.
+	for platform, displayName := range ValidArchDisplayNames {
+		osAndArch := strings.Split(platform, "_")
+		if len(osAndArch) != 2 {
 			continue
 		}
-		// Skip files for invalid platforms.
-		displayName, ok := ValidArchDisplayNames[name[0]]
-		if !ok {
-			continue
+		binary := "evergreen"
+		if strings.Contains(platform, "windows") {
+			binary += ".exe"
 		}
-		osArchParts := strings.Split(name[0], "_")
-		c.ClientBinaries = append(c.ClientBinaries, ClientBinary{
-			URL:         fmt.Sprintf("%s/%s", c.S3URLPrefix, item),
-			OS:          osArchParts[0],
-			Arch:        osArchParts[1],
+
+		clientBinary := ClientBinary{
+			// The items in S3 are expected to be of the form <s3_prefix>/<os>_<arch>/evergreen{.exe}
+			URL:         fmt.Sprintf("%s/%s/%s", s3URLPrefix, platform, binary),
+			OS:          osAndArch[0],
+			Arch:        osAndArch[1],
 			DisplayName: displayName,
-		})
-	}
-	if err = iter.Err(); err != nil {
-		return errors.Wrap(err, "iterating client bucket contents")
+		}
+
+		// Check that the client exists and is accessible.
+		resp, err := client.Head(clientBinary.URL)
+		checkFailedMsg := message.Fields{
+			"message": "could not check for existence of Evergreen client binary for this OS/arch, skipping it and continuing app startup",
+			"os":      clientBinary.OS,
+			"arch":    clientBinary.Arch,
+			"url":     clientBinary.URL,
+		}
+		if err != nil {
+			grip.Notice(message.WrapError(err, checkFailedMsg))
+			continue
+		}
+
+		_ = resp.Body.Close()
+		if resp.StatusCode >= 400 {
+			checkFailedMsg["status_code"] = resp.StatusCode
+			grip.Notice(checkFailedMsg)
```

Comment:
This seems like it'd be expected on staging.

Reply:
It is expected, but that's why it's a notice and not a warning/error. If something goes awry where prod doesn't have a binary though, we can use this to query whether there was an issue with finding all the required binaries.